### PR TITLE
Object Store built with off-chain data

### PR DESCRIPTION
**Pull request by sanderpick**
- OS actor state is now tracking objects that contain pointers to off-chain data, currently hosted in local IPFS. IPFS is used, likely as a temporary solution, because it has built in sync capabilities, and is an off-the-shelf tool for file chunking. The idea is to put this behind an S3 adapter. The IPFS node can be on a private network if we decide to continue down this path. 
- Adds custom handling to resolve Cids from IPFS using a modified copy of the [IPLD resolver](https://github.com/consensus-shipyard/ipc/tree/main/fendermint/vm/resolver) and new methods in the [IPLD resolver service](https://github.com/consensus-shipyard/ipc/tree/main/ipld/resolver). There's more work to do here: Add a voting mechanism to determine when enough nodes have resolved a Cid, modeled after the topdown finality mechanism, and include some proof of storage using FVM randomness in the block preparation step.
- Adds a syscall to unpin Cids when they are deleted from an object store.
- Various cleanups.

### CODE REVIEW

File: ipld/resolver/src/service.rs [hunk 1]
```diff
@@ -500,6 +514,24 @@ where
         }
     }
 
+    /// Start a CID resolution using local IPFS.
+    fn start_ipfs_query(&mut self, cid: Cid, response_channel: ResponseChannel) {
```
Comment (carsonfarmer) [inline ipld/resolver/src/service.rs:518]: ❤️

**Review by carsonfarmer — APPROVED** [review]
Wow. This was a lot of work! I follow most of what is going on here, but some of the _ordering_ of operations isn't 100% clear to me. I'm certainly not going to block this. I have made some comments, but mostly they are either questions or tiny nits that can mostly be ignored.

There is also quite a bit of aesthetic changes in this PR, so I've tried to mostly just ignore those and focus on the core changes.

File: fendermint/actors/accumulator/src/actor.rs [hunk 2]
```diff
@@ -3,18 +3,18 @@
...
-use fil_actors_runtime::actor_dispatch;
-use fil_actors_runtime::actor_error;
-use fil_actors_runtime::builtin::singletons::SYSTEM_ACTOR_ADDR;
-use fil_actors_runtime::runtime::{ActorCode, Runtime};
-use fil_actors_runtime::ActorDowncast;
-use fil_actors_runtime::ActorError;
-// use fvm_ipld_encoding::de::DeserializeOwned;
-// use fvm_ipld_encoding::ser::Serialize;
-use fvm_shared::error::ExitCode;
+use fil_actors_runtime::{
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/actors/accumulator/src/actor.rs:6]: this was a manual tweak... i'm just using rustfmt

File: fendermint/app/config/default.toml [hunk 3]
```diff
@@ -194,7 +194,7 @@ rate_limit_period = 0
 # IPC related configuration parameters
 [ipc]
 # Default subnet ID, which basically means IPC is disabled.
-subnet_id = "/r0"
+subnet_id = "/r31415926"
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/app/config/default.toml:197]: it's from the ipc example docs in the repo, but using it here to force enable the IPLD resolver.

File: fendermint/app/options/src/rpc.rs [hunk 4]
```diff
@@ -101,52 +108,46 @@ pub enum RpcQueryCommands {
...
+}
+
+#[derive(Subcommand, Debug, Clone)]
+pub enum RpcAccumulatorCommands {
+    /// Push a new event into an accumulator.
     Push {
         /// Content to push into the data repo for tracking.
         #[arg(long, short)]
-        content: Bytes,
+        event: Bytes,
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/app/options/src/rpc.rs:148]: cool, we can switch back

File: fendermint/app/src/cmd/run.rs [hunk 5]
```diff
@@ -190,6 +199,19 @@ async fn run(settings: Settings) -> anyhow::Result<()> {
 
         tracing::info!("starting the IPLD Resolver...");
         tokio::spawn(async move { resolver.run().await });
+
+        tracing::info!("starting the IPFS Resolver...");
+        tokio::spawn(async move { ipfs_resolver.run().await });
+
+        // let foo = client
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/app/src/cmd/run.rs:206]: yep, will clean up on next PR :pray:

File: fendermint/vm/interpreter/src/chain.rs [hunk 6]
```diff
@@ -194,6 +222,21 @@ where
                         return Ok(false);
                     }
                 }
+                ChainMessage::Ipc(IpcMessage::ObjectResolved(obj)) => {
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/vm/interpreter/src/chain.rs:225]: yep!

File: fendermint/vm/interpreter/src/chain.rs [hunk 7]
```diff
@@ -231,8 +274,16 @@ where
             ChainMessage::Signed(msg) => {
                 let (state, ret) = self
                     .inner
-                    .deliver(state, VerifiableMessage::Signed(msg))
+                    .deliver(state, VerifiableMessage::Signed(msg.clone()))
                     .await?;
+
+                if ret.is_ok() {
```
[the remark this answers is not in the document]
Comment (sanderpick) [reply fendermint/vm/interpreter/src/chain.rs:280]: yep, we could add our own interpreter to the stack in the future

**Pull request merged**